lib/sdk/tabs/tab-firefox.js

```diff
@@ -3,309 +3,307 @@
-
-  /**
-   * tab's document readyState, or 'uninitialized' if it doesn't even exist yet.
-   */
-  get readyState() {
-    let doc = this._contentDocument;
-    return doc && doc.readyState || 'uninitialized';
+
+  set url(val) {
+    if (isClosed(this))
+      return;
+
+    setTabURL(viewsFor.get(this), val);
   },
 
-  /**
-   * Unique id for the tab, actually maps to tab.linkedPanel but with some munging.
-   */
-  get id() this._tab ? getTabId(this._tab) : undefined,
-
-  /**
-   * The title of the page currently loaded in the tab.
-   * Changing this property changes an actual title.
-   * @type {String}
-   */
-  get title() this._tab ? getTabTitle(this._tab) : undefined,
-  set title(title) this._tab && setTabTitle(this._tab, title),
-
-  /**
-   * Returns the MIME type that the document loaded in the tab is being
-   * rendered as.
-   * @type {String}
-   */
-  get contentType() this._tab ? getTabContentType(this._tab) : undefined,
-
-  /**
-   * Location of the page currently loaded in this tab.
-   * Changing this property will loads page under under the specified location.
-   * @type {String}
-   */
-  get url() this._tab ? getTabURL(this._tab) : undefined,
-  set url(url) this._tab && setTabURL(this._tab, url),
-  /**
-   * URI of the favicon for the page currently loaded in this tab.
-   * @type {String}
-   */
   get favicon() {
     deprecateUsage(
       'tab.favicon is deprecated, ' +
       'please use require("sdk/places/favicon").getFavicon instead.'
     );
-    return this._tab ? getFaviconURIForLocation(this.url) : undefined
+    return isClosed(this) ? undefined : getFaviconURIForLocation(this.url);
+  },
+
+  get contentType() {
+    return isClosed(this) ? undefined : browser(this).documentContentType;
+  },
+
+  get index() {
+    return isClosed(this) ? undefined : getIndex(viewsFor.get(this));
+  },
+
+  set index(val) {
+    if (isClosed(this))
+      return;
+
+    move(viewsFor.get(this), val);
   },
-  /**
-   * The CSS style for the tab
-   */
-  get style() null, // TODO
-  /**
-   * The index of the tab relative to other tabs in the application window.
-   * Changing this property will change order of the actual position of the tab.
-   * @type {Number}
-   */
-  get index()
-    this._tab ?
-    this._window.gBrowser.getBrowserIndexForDocument(this._contentDocument) :
-    undefined,
-  set index(value)
-    this._tab && this._window.gBrowser.moveTabTo(this._tab, value),
-  /**
-   * Thumbnail data URI of the page currently loaded in this tab.
-   * @type {String}
-   */
-  getThumbnail() {
-    if (!this._tab)
+
+  get isPinned() {
+    return isClosed(this) ? undefined : viewsFor.get(this).pinned;
+  },
+
+  get window() {
+    if (isClosed(this))
       return undefined;
-    if (this._tab.getAttribute('remote')) {
-      console.error('This method is not supported with E10S');
-      return BLANK;
-    }
-    return getThumbnailURIForWindow(this._contentWindow);
+
+    require('../windows');
```

Comment:
It creates a circular dependency, that is why we can't have it at the top of the document with all of the other `require` statements.  We need to break the circular dependency.  Probably the easiest way to do this is to define a module which exports both the `Window` constructor and a `Tab` constructor.